src/Umbraco.Infrastructure/PropertyEditors/BlockGridPropertyEditorBase.cs

```diff
@@ -0,0 +1,106 @@
+
+    protected override IDataValueEditor CreateValueEditor() =>
+        DataValueEditorFactory.Create<BlockGridEditorPropertyValueEditor>(Attribute!);
+
+    private class BlockGridEditorPropertyValueEditor : BlockEditorPropertyValueEditor
+    {
+        public BlockGridEditorPropertyValueEditor(
+            DataEditorAttribute attribute,
+            PropertyEditorCollection propertyEditors,
+            IDataTypeService dataTypeService,
+            ILocalizedTextService textService,
+            ILogger<BlockEditorPropertyValueEditor> logger,
+            IShortStringHelper shortStringHelper,
+            IJsonSerializer jsonSerializer,
+            IIOHelper ioHelper,
+            IContentTypeService contentTypeService,
+            IPropertyValidationService propertyValidationService)
+            : base(attribute, propertyEditors, dataTypeService, textService, logger, shortStringHelper, jsonSerializer, ioHelper)
+        {
+            BlockEditorValues = new BlockEditorValues(new BlockGridEditorDataConverter(jsonSerializer), contentTypeService, logger);
+            Validators.Add(new BlockEditorValidator(propertyValidationService, BlockEditorValues, contentTypeService));
+            Validators.Add(new MinMaxValidator(BlockEditorValues, textService));
+        }
+
+        private class MinMaxValidator : BlockEditorMinMaxValidatorBase
+        {
+            private readonly BlockEditorValues _blockEditorValues;
+
+            public MinMaxValidator(BlockEditorValues blockEditorValues, ILocalizedTextService textService)
+                : base(textService) =>
+                _blockEditorValues = blockEditorValues;
+
+            public override IEnumerable<ValidationResult> Validate(object? value, string? valueType, object? dataTypeConfiguration)
+            {
+                if (dataTypeConfiguration is not BlockGridConfiguration blockConfig)
+                {
+                    return Array.Empty<ValidationResult>();
+                }
+
+                BlockEditorData? blockEditorData = _blockEditorValues.DeserializeAndClean(value);
+
+                var validationResults = new List<ValidationResult>();
+                validationResults.AddRange(ValidateNumberOfBlocks(blockEditorData, blockConfig.ValidationLimit.Min, blockConfig.ValidationLimit.Max));
+
+                var areasConfigsByKey = blockConfig.Blocks.SelectMany(b => b.Areas).ToDictionary(a => a.Key);
+
+                IList<BlockGridLayoutAreaItem> ExtractLayoutAreaItems(BlockGridLayoutItem item)
+                {
+                    var areas = item.Areas.ToList();
+                    areas.AddRange(item.Areas.SelectMany(a => a.Items).SelectMany(ExtractLayoutAreaItems));
+                    return areas;
+                }
+
+                BlockGridLayoutAreaItem[]? areas = blockEditorData?.Layout?.ToObject<IEnumerable<BlockGridLayoutItem>>()?.SelectMany(ExtractLayoutAreaItems).ToArray();
+
+                if (areas?.Any() != true)
```

Comment:
It should be a bit faster to used `Length` instead of any since it already know the size of the array, while `Any()` iterate first element to check if there is any items (at least one) in the collection.

Reply:
It's not completely true.. Any() checks if the type implements `ICollection` and if so, it uses the known count.

There of cause is an overhead, but I'm not sure if it is worth if, compared to readability.

Below is the implementation of `Any()` from `System.Linq`

```cs
public static bool Any<TSource>(this IEnumerable<TSource> source)
{
    if (source == null)
    {
        ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
    }

    if (source is ICollection<TSource> collectionoft)
    {
        return collectionoft.Count != 0;
    }
    else if (source is IIListProvider<TSource> listProv)
    {
        // Note that this check differs from the corresponding check in
        // Count (whereas otherwise this method parallels it).  If the count
        // can't be retrieved cheaply, that likely means we'd need to iterate
        // through the entire sequence in order to get the count, and in that
        // case, we'll generally be better off falling through to the logic
        // below that only enumerates at most a single element.
        int count = listProv.GetCount(onlyIfCheap: true);
        if (count >= 0)
        {
            return count != 0;
        }
    }
    else if (source is ICollection collection)
    {
        return collection.Count != 0;
    }

    using (IEnumerator<TSource> e = source.GetEnumerator())
    {
        return e.MoveNext();
    }
}
```

Reply:
@bergmania okay, I think I have seende a similar explanation before regarding collections of type `IList`, `ICollection` etc.. but generally with enumerable collections it would perform a letter better.

Personally I prefer using `Length` or `Count` properties when I need to check whether there's no elements or at least one element.

Reply:
Also remember that `areas?.Length == 0` fails if `areas` is `null`, where `areas?.Any() != true` succeeds for `areas` being both `null` and empty. The equivalent using `Length` would be `areas == null || areas.Length == 0`, and I do belive that `areas?.Any() != true` is better for readability.

Reply:
@kjac yes, you're right. Mostly I use the `?.Count > 0` approach.

Some implements `HasAny()` or `IsEmpty()` extention method to handle both null and empty collections, but it may collide with extension methods in other libraries, so not sure if it makes things simpler in Umbraco core.